include/permanent/ryser.h

```diff
@@ -95,120 +126,72 @@ namespace permanent {
 template <typename T, typename I = void>
 result_t<T, I> ryser_square(const size_t m, const size_t n, const T *ptr)
 {
-  (void)n;
-
-  size_t i, j;
-  size_t k;
-  result_t<T, I> rowsum;
-  result_t<T, I> out = 0;
-
-  /* Iterate over c = pow(2, m) submatrices (equal to (1 << m)) submatrices.
-   */
-
-  size_t c = 1UL << m;
-
-  /* Loop over columns of submatrix; compute product of row sums. */
-
-  for (k = 0; k < c; ++k) {
-    result_t<T, I> rowsumprod = 1.0;
-    for (i = 0; i < m; ++i) {
-      /* Loop over rows of submatrix; compute row sum. */
-
-      rowsum = 0.0;
-      for (j = 0; j < m; ++j) {
-        /* Add element to row sum if the row index is in the
-         * characteristic * vector of the submatrix, which is the binary
-         * vector given by k.  */
-
-        if (k & (1UL << j)) {
-          rowsum += ptr[m * i + j];
+    (void)n;
+
+    size_t i, j, k;
+    T rowsum;  // Use T instead of result_t for intermediate calculations
+    T out = 0;
+    size_t c = 1UL << m;
+
+    for (k = 0; k < c; ++k) {
+        T rowsumprod = 1.0;
+        for (i = 0; i < m; ++i) {
+            rowsum = 0.0;
+            for (j = 0; j < m; ++j) {
+                if (k & (1UL << j)) {
+                    rowsum += ptr[m * i + j];
+                }
+            }
+            rowsumprod *= rowsum;
         }
-      }
-
-      /* Update product of row sums. */
-
-      rowsumprod *= rowsum;
+        out += rowsumprod * (1 - ((__builtin_popcountll(k) & 1) << 1));
```

Comment:
Use std::popcount here too.  It's new in C++20, which we're using... it's best to avoid these compiler extensions if possible.